_posts/2016-09-28-when-a-good-architecture-helps-you-in-your-ui-design-iterations.md

```diff
@@ -0,0 +1,104 @@
+canonical:
+    name: my personal blog
+    href: http://www.ananogal.com/blog/when-a-good-architecture-helps-you-in-your-ui-design-iterations/
+image:
+    src: /assets/img/custom/blog/2016-09-28-memento-iterations.png
+
+tags:
+- Swift
+- Memento
+- VIPER
+- Separation of concerns
+- UI design
+- Clean architecture
+
+---
+I’m currently working on an internal app for Codurance, which started as a pet project while I was working with my previous client. I started the project with one simple goal: to make it work. After six or seven months of trying to put it all together in my spare time, I ended up with a skeleton of a working app. However, I wasn't entirely satisfied with my work. Even though I had been test-driven all the code and all my ViewControllers were as light as I could imagine, I thought it wasn't enough.
+So I decided to carry out a big refactoring and apply the VIPER architectural pattern.
+
+## So what's VIPER?
+
+VIPER stands for View, Interactor, Presenter, Entity, and Router, and it's a version of the [clean architecture](https://8thlight.com/blog/uncle-bob/2012/08/13/the-clean-architecture.html) applied for iOS.
+It's not my intention to take a deep-dive into VIPER in this article, but if you are interested in learning more, [here](https://www.objc.io/issues/13-architecture/viper/), you have a good definition of the architecture, and [here](https://swifting.io/blog/2016/03/07/8-viper-to-be-or-not-to-be/) you can find a good explanation that can help you decide if you want to use it in your app or not. The section "Benefits of VIPER" in [this article](https://brigade.engineering/brigades-experience-using-an-mvc-alternative-36ef1601a41f#.d8zy2b507) explain quite well why I decided to use VIPER.
+
+## Make it work
+So as I explained, originally my main objective had been to make it work. And I did. With my big refactoring in place, I finished my goals for the app in no time, having the major features working. The app looked like this at that time:
+
+<img src="/assets/img/custom/blog/2016-09-28-memento-first-version.png" alt="first version" width="169" height="300" class="image-side-by-side-text" />
+
+<p class="text-side-by-side-image" >
+<br/>
+This version had several problems: The way to add or delete new TimeEntries was completely obscure for a new user: you had to swipe the cell to see the buttons. I was using this pod: <a href ="https://github.com/CEWendel/SWTableViewCell" >SWTableViewCell</a>  to achieve this.
+I was happy because my app was working, but I was sad because it wasn't an app that I would like to show off. So I had a talk with our <a href ="https://twitter.com/GiuliaMantuano" >UI expert</a>, and we decided to start iterating till we get the UI into a more satisfactory state.
+
+</p>
+<br class="clear-both" />
+
+## First Iteration: Copy Google Calendar
+
+<img src="/assets/img/custom/blog/2016-09-28-memento-second-version.png" alt="second version" width="169" height="300" class="image-side-by-side-text" />
```

Comment:
All the images are quite small. Any chance you could crop them to make it easier to read the buttons and links in the UI?